<a href="https://colab.research.google.com/github/Huwalli/PLta/blob/main/%E3%80%8CExample2_DataClearning_ipynb%E3%80%8D%E7%9A%84%E5%89%AF%E6%9C%AC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

金門縣政府提供教育相關數據

https://drive.google.com/drive/folders/1ZnEmLN_Gbx074-dv_8m3chn2vvhaeUSK?usp=sharing

金門縣康軒國中小使用數據

https://docs.google.com/spreadsheets/d/1LvcLpeVNIa-OfICFnakSbzQgSKBCkfFACmQRIKekgRA/edit?usp=sharing

In [ ]:
from google.colab import auth
auth.authenticate_user()

import gspread
from google.auth import default
creds, _ = default()

gc = gspread.authorize(creds)

In [ ]:
import pandas as pd
# read data and put it in a dataframe
# 在 google 工作表載入 gsheets
gsheets = gc.open_by_url('https://docs.google.com/spreadsheets/d/1V4c0lHTNIfSW3PX52pL4AkGCQ7x9YX3oqWf6pVlyXYA/edit?usp=sharing')

In [ ]:
# 從 gsheets 的 All-whiteboard-device 載入 sheets
sheets = gsheets.worksheet('anonymized_xapi').get_all_values()
# 將 sheets1 資料載入 pd 的 DataFrame 進行分析
df = pd.DataFrame(sheets[1:], columns=sheets[0])
# 取得最前面的5筆資料
df.head()

id                                               json      timestamp
0  281744  {"id": "5d3ddc37-c43c-4342-8015-c1697ac5148b",...  6/11/24 22:17
1  281746  {"id": "b27139a8-101a-4d94-9f24-0fb6be429de7",...  6/11/24 22:17
2  281747  {"id": "73d06d71-56b5-4fd2-a951-7f3e485294ba",...  6/11/24 22:17
3  281748  {"id": "3c0c29d1-c8f9-441f-a820-bafd90e0fc58",...  6/11/24 22:17
4  281749  {"id": "6cce33e0-b8d4-47f4-a00d-6dd335c409c2",...  6/11/24 22:17

In [4]:
# 設定alldays為時間戳最大-最小，也就是資料調查天數
df['timestamp'] = pd.to_datetime(df['timestamp'])
alldays = max(set(df['timestamp'].dt.date)) - min(set(df['timestamp'].dt.date))

<ipython-input-4-caecfaad91b4>:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['timestamp'] = pd.to_datetime(df['timestamp'])


In [5]:
alldays.days

37

In [6]:
len(df)

42928

In [21]:
# 將json檔展開成表格
import json
df_expanded = pd.json_normalize(df['json'].apply(json.loads))

In [23]:
df_expanded

id                      timestamp  \
0      5d3ddc37-c43c-4342-8015-c1697ac5148b  2024-06-11T22:17:50.069+08:00   
1      b27139a8-101a-4d94-9f24-0fb6be429de7  2024-06-11T22:17:53.288+08:00   
2      73d06d71-56b5-4fd2-a951-7f3e485294ba  2024-06-11T22:17:54.186+08:00   
3      3c0c29d1-c8f9-441f-a820-bafd90e0fc58  2024-06-11T22:17:55.819+08:00   
4      6cce33e0-b8d4-47f4-a00d-6dd335c409c2  2024-06-11T22:17:56.314+08:00   
...                                     ...                            ...   
42923  daded534-94e8-4227-b88b-4cff5ba805c8  2024-07-18T11:24:45.780+08:00   
42924  194b1d8f-fdf9-48f1-9093-23b0bd33c536  2024-07-18T11:26:18.140+08:00   
42925  2ab53b3f-ab35-4728-bc0b-1a098ddae7d5  2024-07-18T11:26:24.491+08:00   
42926  481ea062-d731-4455-a668-2802558d8faf  2024-07-18T11:28:08.280+08:00   
42927  2fbbda86-8053-44cd-b7d6-0fef8eb9e229  2024-07-18T11:29:13.434+08:00   

      actor.objectType actor.name       actor.account.homePage  \
0                Agent      學生298  mailto:kc0368@cnc.km.edu.tw   
1                Agent      學生298  mailto:kc0368@cnc.km.edu.tw   
2                Agent      學生298  mailto:kc0368@cnc.km.edu.tw   
3                Agent      學生298  mailto:kc0368@cnc.km.edu.tw   
4                Agent      學生298  mailto:kc0368@cnc.km.edu.tw   
...                ...        ...                          ...   
42923            Agent      學生584  mailto:kd8888@cnc.km.edu.tw   
42924            Agent      學生584  mailto:kd8888@cnc.km.edu.tw   
42925            Agent      學生584  mailto:kd8888@cnc.km.edu.tw   
42926            Agent      學生584  mailto:kd8888@cnc.km.edu.tw   
42927            Agent      學生584  mailto:kd8888@cnc.km.edu.tw   

      actor.account.name                                      verb.id  \
0                     學生  https://w3id.org/xapi/acrossx/verbs/watched   
1                     學生   https://w3id.org/xapi/acrossx/verbs/paused   
2                     學生  https://w3id.org/xapi/acrossx/verbs/watched   
3                     學生   https://w3id.org/xapi/acrossx/verbs/paused   
4                     學生  https://w3id.org/xapi/acrossx/verbs/watched   
...                  ...                                          ...   
42923                 學生   https://w3id.org/xapi/acrossx/verbs/paused   
42924                 學生  https://w3id.org/xapi/acrossx/verbs/watched   
42925                 學生   https://w3id.org/xapi/acrossx/verbs/paused   
42926                 學生  https://w3id.org/xapi/acrossx/verbs/watched   
42927                 學生   https://w3id.org/xapi/acrossx/verbs/paused   

      verb.display.zh-TW object.objectType  \
0                     觀看          Activity   
1                     暫停          Activity   
2                     觀看          Activity   
3                     暫停          Activity   
4                     觀看          Activity   
...                  ...               ...   
42923                 暫停          Activity   
42924                 觀看          Activity   
42925                 暫停          Activity   
42926                 觀看          Activity   
42927                 暫停          Activity   

                                               object.id  ...  \
0      https://ksjh-km.kschool.com.tw/video-player/44...  ...   
1      https://ksjh-km.kschool.com.tw/video-player/44...  ...   
2      https://ksjh-km.kschool.com.tw/video-player/44...  ...   
3      https://ksjh-km.kschool.com.tw/video-player/44...  ...   
4      https://ksjh-km.kschool.com.tw/video-player/44...  ...   
...                                                  ...  ...   
42923  https://khjh-km.kschool.com.tw/video-player/51...  ...   
42924  https://khjh-km.kschool.com.tw/video-player/51...  ...   
42925  https://khjh-km.kschool.com.tw/video-player/51...  ...   
42926  https://khjh-km.kschool.com.tw/video-player/51...  ...   
42927  https://khjh-km.kschool.com.tw/video-player/51...  ...   

      context.team.name        context.platform context.team.objectType  \
0                714504  

In [14]:
# 資料格式
type(df_expanded['timestamp'][0])

str

In [24]:
df_expanded['id'] = df['id']  # 加入每筆紀錄的 id（用來識別每一筆行為）
df_expanded['timestamp'] = pd.to_datetime(df['timestamp'])  # 將時間欄位轉換成「時間格式」，這樣才能進一步分析時間（例如：幾月幾號、幾點）
df_expanded['Date'] = df_expanded['timestamp'].dt.date   # 從時間中提取出「日期」部分，例如 2024-06-11
df_expanded['Time'] = df_expanded['timestamp'].dt.time    # 從時間中提取出「時間」部分，例如 22:17:50
df_expanded['Weekday'] = df_expanded['timestamp'].dt.day_name()  # 取得星期幾的名稱，例如 Monday、Tuesday，方便看行為是發生在哪一天
df_expanded.head()  # 顯示處理後的前五筆資料，幫助我們確認資料長得正不正確


id           timestamp actor.objectType actor.name  \
0  281744 2024-06-11 22:17:00            Agent      學生298   
1  281746 2024-06-11 22:17:00            Agent      學生298   
2  281747 2024-06-11 22:17:00            Agent      學生298   
3  281748 2024-06-11 22:17:00            Agent      學生298   
4  281749 2024-06-11 22:17:00            Agent      學生298   

        actor.account.homePage actor.account.name  \
0  mailto:kc0368@cnc.km.edu.tw                 學生   
1  mailto:kc0368@cnc.km.edu.tw                 學生   
2  mailto:kc0368@cnc.km.edu.tw                 學生   
3  mailto:kc0368@cnc.km.edu.tw                 學生   
4  mailto:kc0368@cnc.km.edu.tw                 學生   

                                       verb.id verb.display.zh-TW  \
0  https://w3id.org/xapi/acrossx/verbs/watched                 觀看   
1   https://w3id.org/xapi/acrossx/verbs/paused                 暫停   
2  https://w3id.org/xapi/acrossx/verbs/watched                 觀看   
3   https://w3id.org/xapi/acrossx/verbs/paused                 暫停   
4  https://w3id.org/xapi/acrossx/verbs/watched                 觀看   

  object.objectType                                          object.id  ...  \
0          Activity  https://ksjh-km.kschool.com.tw/video-player/44...  ...   
1          Activity  https://ksjh-km.kschool.com.tw/video-player/44...  ...   
2          Activity  https://ksjh-km.kschool.com.tw/video-player/44...  ...   
3          Activity  https://ksjh-km.kschool.com.tw/video-player/44...  ...   
4          Activity  https://ksjh-km.kschool.com.tw/video-player/44...  ...   

  context.team.mailto context.team.mbox  \
0                 NaN               NaN   
1                 NaN               NaN   
2                 NaN               NaN   
3                 NaN               NaN   
4                 NaN               NaN   

  object.definition.extensions.https://w3id.org/xapi/acrossx/extensions/alignment  \
0                                                NaN                                
1                                                NaN                                
2                                                NaN                                
3                                                NaN                                
4                                                NaN                                

  object.definition.name.zh-TW object.definition.interactionType  \
0                          NaN                               NaN   
1                          NaN                               NaN   
2                          NaN                               NaN   
3                          NaN                               NaN   
4                          NaN                               NaN   

  object.definition.correctResponsesPattern object.definition.choices  \
0                                       NaN                       NaN   
1                                       NaN                       NaN   
2                                       NaN                       NaN   
3                                       NaN                       NaN   
4                                       NaN                       NaN   

         Date      Time  Weekday  
0  2024-06-11  22:17:00  Tuesday  
1  2024-06-11  22:17:00  Tuesday  
2  2024-06-11  22:17:00  Tuesday  
3  2024-06-11  22:17:00  Tuesday  
4  2024-06-11  22:17:00  Tuesday  

[5 rows x 25 columns]

In [25]:
# 根據「學生名稱」和「日期」把資料分組（每個學生每天一組）
grouped = df_expanded.groupby(['actor.name', 'Date']).agg({

    # 把每一天的時間戳記都收集起來，變成一個列表
    'timestamp': list,

    # 把每一天學生做過的動作（例如：觀看、回答）都收集成一個列表
    'verb.display.zh-TW': list,

    # 把每一天互動過的學習資源（例如影片、題目）的 ID 收集起來
    'object.id': list

# 聚合完資料後，重設索引，把分組的欄位變成正常欄位
}).reset_index()


In [26]:
# 把「日期」的資料格式轉換成程式能理解的時間格式（方便後面做處理）
grouped['Date'] = pd.to_datetime(grouped['Date'])

# 從日期中找出是星期幾，新增一欄叫做「Weekday」（例如：Monday、Tuesday）
grouped['Weekday'] = grouped['Date'].dt.day_name()
grouped

actor.name       Date                                          timestamp  \
0        學生298 2024-06-11  [2024-06-11 22:17:00, 2024-06-11 22:17:00, 202...   
1        學生298 2024-06-12  [2024-06-12 11:39:00, 2024-06-12 11:46:00, 202...   
2        學生298 2024-06-17  [2024-06-17 13:39:00, 2024-06-17 13:48:00, 202...   
3        學生298 2024-06-19  [2024-06-19 11:16:00, 2024-06-19 11:51:00, 202...   
4        學生298 2024-06-24  [2024-06-24 13:47:00, 2024-06-24 14:04:00, 202...   
..         ...        ...                                                ...   
868      學生590 2024-06-25  [2024-06-25 13:06:00, 2024-06-25 13:07:00, 202...   
869      學生591 2024-06-25  [2024-06-25 13:07:00, 2024-06-25 13:09:00, 202...   
870      學生592 2024-06-28         [2024-06-28 15:09:00, 2024-06-28 15:10:00]   
871      學生593 2024-07-02                              [2024-07-02 11:21:00]   
872      學生594 2024-07-02         [2024-07-02 19:44:00, 2024-07-02 19:47:00]   

                                    verb.display.zh-TW  \
0                       [觀看, 暫停, 觀看, 暫停, 觀看, 暫停, 開始嘗試]   
1    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
2    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
3    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
4    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
..                                                 ...   
868  [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
869  [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
870                                           [查看, 查看]   
871                                               [查看]   
872                                           [觀看, 暫停]   

                                             object.id    Weekday  
0    [https://ksjh-km.kschool.com.tw/video-player/4...    Tuesday  
1    [https://ksjh-km.kschool.com.tw/student-quiz-a...  Wednesday  
2    [https://ksjh-km.kschool.com.tw/student-quiz-a...     Monday  
3    [https://ksjh-km.kschool.com.tw/student-quiz-a...  Wednesday  
4    [https://ksjh-km.kschool.com.tw/student-quiz-a...     Monday  
..                                                 ...        ...  
868  [https://saps-km.kschool.com.tw/student-quiz-a...    Tuesday  
869  [https://saps-km.kschool.com.tw/student-quiz-a...    Tuesday  
870  [https://saps-km.kschool.com.tw/teacher-readin...     Friday  
871  [https://ksjh-km.kschool.com.tw/teacher-readin...    Tuesday  
872  [https://gnes-km.kschool.com.tw/video-player/1...    Tuesday  

[873 rows x 6 columns]

In [27]:
len(set(grouped['actor.name'])) #有297位學生

297

根據每位學生的瀏覽日期，來去畫出不同學生瀏覽影片的天數

In [32]:
# 計算每個學生（actor.name）有「幾天」有活動紀錄（用 nunique() 計算不重複的天數）
student_days = grouped.groupby('actor.name')['Date'].nunique()

# 嘗試幫結果命名欄位
student_days.colnames = ['actor.name', 'Date']

# 把索引轉回欄位，讓結果變成一個普通的表格
student_days = student_days.reset_index()
student_days

actor.name  Date
0        學生298     5
1        學生299     9
2        學生300    11
3        學生301     8
4        學生302    11
..         ...   ...
292      學生590     1
293      學生591     1
294      學生592     1
295      學生593     1
296      學生594     1

[297 rows x 2 columns]

In [34]:
student_days[student_days['Date'] > (alldays.days / 4)]['actor.name']

2     學生300
4     學生302
5     學生303
6     學生304
7     學生305
9     學生307
11    學生309
13    學生311
14    學生312
15    學生313
16    學生314
17    學生315
97    學生395
Name: actor.name, dtype: object

In [36]:
import plotly.express as px
# 計算每個 Date 值出現多少人
date_counts = student_days['Date'].value_counts().reset_index()
date_counts.columns = ['Date', 'Count']

# 使用 Plotly 繪製條形圖
fig = px.bar(date_counts.sort_values(by='Date'),
             x='Date', y='Count',
             labels={'Date': '天數', 'Count': '人數'},
             title='每個天數對應的人數')

# 顯示圖表
fig.show()

In [37]:
# 選出「名字是 學生315」的所有資料（也就是只看學生315的行為紀錄）
grouped[grouped['actor.name'] == '學生315']

actor.name       Date                                          timestamp  \
164      學生315 2024-06-12  [2024-06-12 08:40:00, 2024-06-12 08:57:00, 202...   
165      學生315 2024-06-13  [2024-06-13 12:25:00, 2024-06-13 12:55:00, 202...   
166      學生315 2024-06-14  [2024-06-14 12:24:00, 2024-06-14 12:36:00, 202...   
167      學生315 2024-06-17  [2024-06-17 12:27:00, 2024-06-17 12:59:00, 202...   
168      學生315 2024-06-18  [2024-06-18 12:28:00, 2024-06-18 12:53:00, 202...   
169      學生315 2024-06-19  [2024-06-19 07:26:00, 2024-06-19 07:43:00, 202...   
170      學生315 2024-06-21  [2024-06-21 12:30:00, 2024-06-21 13:10:00, 202...   
171      學生315 2024-06-23  [2024-06-23 15:47:00, 2024-06-23 15:47:00, 202...   
172      學生315 2024-06-24  [2024-06-24 12:31:00, 2024-06-24 13:12:00, 202...   
173      學生315 2024-06-25  [2024-06-25 12:24:00, 2024-06-25 12:50:00, 202...   
174      學生315 2024-06-26  [2024-06-26 12:30:00, 2024-06-26 12:36:00, 202...   
175      學生315 2024-06-27  [2024-06-27 12:23:00, 2024-06-27 12:42:00, 202...   

                                    verb.display.zh-TW  \
164  [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
165  [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
166  [開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回...   
167  [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
168  [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
169  [開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回...   
170  [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
171  [開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回...   
172  [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   
173  [開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 回答, 回答, 回...   
174  [開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回...   
175  [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...   

                                             object.id    Weekday  
164  [https://ksjh-km.kschool.com.tw/student-quiz-a...  Wednesday  
165  [https://ksjh-km.kschool.com.tw/student-quiz-a...   Thursday  
166  [https://ksjh-km.kschool.com.tw/student-quiz-a...     Friday  
167  [https://ksjh-km.kschool.com.tw/student-quiz-a...     Monday  
168  [https://ksjh-km.kschool.com.tw/student-quiz-a...    Tuesday  
169  [https://ksjh-km.kschool.com.tw/student-quiz-a...  Wednesday  
170  [https://ksjh-km.kschool.com.tw/student-quiz-a...     Friday  
171  [https://ksjh-km.kschool.com.tw/student-quiz-a...     Sunday  
172  [https://ksjh-km.kschool.com.tw/student-quiz-a...     Monday  
173  [https://ksjh-km.kschool.com.tw/student-quiz-a...    Tuesday  
174  [https://ksjh-km.kschool.com.tw/student-quiz-a...  Wednesday  
175  [https://ksjh-km.kschool.com.tw/student-quiz-a...   Thursday

In [ ]:
# 只看學生315的行為
grouped[grouped['actor.name'] == '學生315']['verb.display.zh-TW']

164    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...
165    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...
166    [開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回...
167    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...
168    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...
169    [開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回...
170    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...
171    [開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回...
172    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...
173    [開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 回答, 回答, 回...
174    [開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回...
175    [開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答,...
Name: verb.display.zh-TW, dtype: object

In [38]:
# 只看「學生315」做過哪些動作，並統計每個動作出現幾次（像是觀看幾次、回答幾次）
grouped[grouped['actor.name'] == '學生315']['verb.display.zh-TW'].value_counts()


verb.display.zh-TW
[開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成]                                                                                                                                                                                                                                                               4
[開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成]                                                                                     2
[開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成]                                                                                                                                                                                                                                                                                                                           1
[開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成]                                                                                                                                                                                                                                             1
[開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成, 開始嘗試, 開始嘗試, 回答, 回答, 回答, ...]          1
[開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成]                                                                                                                                                                                                                                                         1
[開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, ...]    1
[開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成, 開始嘗試, 開始嘗試, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 回答, 完成]                                                             1
Name: count, dtype: int64

查看同學不同的行為，他們總共有哪些動作

In [39]:
# 把每一天每位學生做過的動作去除重複，只保留「不重複」的動作（轉成 set 集合）
grouped['unique_verbs'] = grouped['verb.display.zh-TW'].apply(lambda x: set(x))

# 把所有學生每天的「不重複動作」打開攤平（explode），變成一個一個的動作列
unique_types = grouped['unique_verbs'].explode()

# 再次轉成 set（集合）來移除重複，得到「所有學生一共出現過哪些動作」
distinct_types = set(unique_types)

distinct_types

{'回答', '完成', '暫停', '查看', '觀看', '開始嘗試'}

透過這個程式碼我們可以調閱不同學生他們不同動作的記數

In [46]:
# 先挑出學生315的資料
student = grouped[grouped['actor.name'] == '學生315']

# 把他所有的動作串起來（是 list of list，要攤平成一個大 list）
all_verbs = student['verb.display.zh-TW'].explode()

# 計算每個動作的出現次數
verb_counts = all_verbs.value_counts()

# 顯示結果
verb_counts


verb.display.zh-TW
回答      798
開始嘗試     33
完成       20
Name: count, dtype: int64

## 想想看，如果我想知道學生314的學習狀況，該怎麼分析，程式碼該怎麼寫?

在上述的數據清理與分析過程中，我們利用了來自金門縣康軒國中小的匿名學習行為數據，並針對學生使用平台的時間與行為進行了探討。

首先，透過 Google Sheets API，我們將資料載入 Pandas DataFrame 進行清理和格式化，進一步將資料轉換為標準的時間戳格式，並提取了其中的日期、時間及星期幾等資訊，以利後續的時間序列分析。

我們對每位學生的行為進行了聚合，計算他們在不同日期進行的操作次數，並分析了學生使用平台的行為動詞（如“回答”、“完成”、“暫停”、“查看”、“觀看”、“開始嘗試”），這些行為動詞描述了學生在平台上的具體學習活動。

根據資料，我們發現共有 297 名學生參與了平台使用，並根據每位學生的瀏覽日期，計算出學生在不同天數瀏覽影片的頻率。

進一步的分析中，我們計算了每位學生在所有天數內進行學習活動的總天數，並篩選出超過全體總天數的四分之一的學生，找出了持續活躍於平台的學生群體。此分析提供了學生學習參與的長期趨勢。

最後，透過 Plotly 繪製的條形圖，我們視覺化了每個天數中使用平台的人數分佈，顯示不同天數對應的活躍學生數量，以幫助教育機構進一步了解學生的學習習慣與學習高峰期。

這些結果可以為教育管理者提供有價值的洞見，幫助他們理解學生的使用情況，並進一步優化平台內容與功能，以提升學生的學習體驗和成效。

在這段數據分析中，我們發現了幾個值得注意的現象：

    學生活躍度的分佈：根據每位學生在不同日期的瀏覽紀錄，我們發現總共有 297 名學生使用該學習平台。但並非所有學生的使用頻率都相同，部分學生的活躍度明顯高於其他人。例如，篩選後發現有些學生的使用頻率超過了所有天數的四分之一，這意味著這些學生在平台上有著較為穩定和頻繁的學習行為。

    學習行為的多樣性：透過分析學生的學習動作（例如“回答”、“完成”、“暫停”、“查看”、“觀看”、“開始嘗試”），我們觀察到學生在使用平台時，進行了多樣化的操作行為，這代表學生在平台上的學習過程不是單一行為，而是存在多階段的互動（如觀看影片、暫停、回答問題等）。這樣的多階段學習行為可能代表學生的學習過程比較深入，而不是僅僅瀏覽或觀看影片。

    學習參與的時間分佈：在分析了學生學習活動的天數後，我們發現不同學生的學習參與天數有明顯差異。有些學生僅在少數天數內使用平台，而另一些學生則在較長的時間內保持活躍。這可能暗示部分學生只在特定時間或事件（例如考試或作業期限）前集中使用平台，而其他學生則更可能是持續學習。

    高峰期現象：從條形圖中，我們能夠清楚看到每個天數對應的人數分佈，這可以幫助我們發現學生使用平台的高峰期。這些高峰期很可能與學校活動、考試或其他學習相關的節點有關。例如，當天數中有明顯的活躍人數激增，可能反映出課程或作業的截止日期，這對於管理者而言是有用的資訊，可以幫助他們在適當時機提供更多學習資源或支援。

這些現象為教育機構提供了學生使用平台的行為模式，有助於針對不同的學習習慣做出相應調整，進一步提高學生的學習效果。

如果是你，要怎麼運用這筆資料 ? 或者你會想要用什麼json資料來分析 ?